File: assets/VAULT_API/Components/Rigidbody3D.cs
using System;
using System.Runtime.CompilerServices;
using Vault;

namespace Vault
{
    public class Rigidbody3D : Component
    {
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_GetKey(string key, string id, out string result);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_SetKey(string key, string id, string value);

        // Component Functions
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_AddForce(string id, float x, float y, float z);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_AddTorque(string id, float x, float y, float z);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_AddForceAtPosition(string id, float fx, float fy, float fz, float px, float py, float pz);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_SetVelocity(string id, float x, float y, float z);
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern internal static void Rigidbody3D_SetAngularVelocity(string id, float x, float y, float z);

        void AddForce(Vector3 force)
        {
            Rigidbody3D_AddForce(Entity.ID, force.x, force.y, force.z);
        }

        void AddTorque(Vector3 torque)
        {
            Rigidbody3D_AddTorque(Entity.ID, torque.x, torque.y, torque.z);
        }

        void AddForceAtPosition(Vector3 force, Vector3 position)
        {
            Rigidbody3D_AddForceAtPosition(Entity.ID, force.x, force.y, force.z, position.x, position.y, position.z);
        }

        void SetVelocity(Vector3 velocity)
        {
            Rigidbody3D_SetVelocity(Entity.ID, velocity.x, velocity.y, velocity.z);
        }

        void SetAngularVelocity(Vector3 velocity)
        {
            Rigidbody3D_SetAngularVelocity(Entity.ID, velocity.x, velocity.y, velocity.z);
        }

        public float mass
        {
            get
            {
                Rigidbody3D_GetKey("mass", Entity.ID, out string result);
                return float.Parse(result);
            }
            set
            {
                Rigidbody3D_SetKey("mass", Entity.ID, Format.ToString(value));
            }
        }

        public float restitution
        {
            get
            {
                Rigidbody3D_GetKey("restitution", Entity.ID, out string result);
                return float.Parse(result);
            }
            set
            {
                Rigidbody3D_SetKey("restitution", Entity.ID, Format.ToString(value));
            }
        }

        public float friction
        {
            get
            {
                Rigidbody3D_GetKey("friction", Entity.ID, out string result);
                return float.Parse(result);
            }
            set
            {
                Rigidbody3D_SetKey("friction", Entity.ID, Format.ToString(value));
            }
        }

        public bool trigger
        {
            get
            {
                Rigidbody3D_GetKey("trigger", Entity.ID, out string result);
                return result == "true";
            }
            set
            {
                Rigidbody3D_SetKey("trigger", Entity.ID, value ? "true" : "false");
            }
        }

        public bool fixedRotation
        {
            get
            {
                Rigidbody3D_GetKey("fixedRotation", Entity.ID, out string result);
                return result == "true";
            }
            set
            {
                Rigidbody3D_SetKey("fixedRotation", Entity.ID, value ? "true" : "false");
            }
        }

        public Vector3 velocity
        {
            get
            {
                Rigidbody3D_GetKey("vecloity", Entity.ID, out string result);
                string[] split = result.Split(" ");
                return new Vector3(float.Parse(split[0]), float.Parse(split[1]), float.Parse(split[2]));
            }
            set
            {
                Rigidbody3D_SetVelocity(Entity.ID, value.x, value.y, value.z);
            }
        }

        public Vector3 angularVelocity
        {
            get
            {
                Rigidbody3D_GetKey("angularVecloity", Entity.ID, out string result);
                string[] split = result.Split(" ");
                return new Vector3(float.Parse(split[0]), float.Parse(split[1]), float.Parse(split[2]));
            }
            set
            {
                Rigidbody3D_SetAngularVelocity(Entity.ID, value.x, value.y, value.z);
            }
        }
    }
}

File: engine/src/Components/Rigidbody3D.cpp
#include "Engine/DecomposeTransform.hpp"
#include "Engine/Physics/BulletPhysics.hpp"
#include "Engine/Runtime.hpp"
#include <Engine/Components/Rigidbody3D.hpp>
#include <Editor/GUI/MainGUI.hpp>
#include <icons/icons.h>

namespace Engine::Components {
    void Rigidbody3D::Init() {
        transform = &Scene::Main->EntityRegistry.get<Transform>(entity);
    }

    void Rigidbody3D::OnGUI() {
        DrawComponent<Rigidbody3D>(Scene::Main->EntityRegistry, 40, [&] {
            ImGui::DragFloat("Mass", &mass, 0.01f);
            ImGui::DragFloat("Friction", &friction, 0.01f);
            ImGui::DragFloat("Restitution", &restitution, 0.01f);
            ImGui::Checkbox("Trigger", &trigger);
            ImGui::Checkbox("Fixed Rotation", &fixedRotation);
        });
    }

    void Rigidbody3D::CreateBody(btCollisionShape *shape) {
        ref = shape;
        bt_transform = new btTransform();
        bt_transform->setIdentity();
        bt_transform->setOrigin(btVector3(transform->position.x,
                                          transform->position.y,
                                          transform->position.z));
        glm::vec3 q = transform->rotation;
        btQuaternion rotation;
        rotation.setEulerZYX(q.z, q.y, q.x);
        std::cout << "Created a Bulet Physics Rigidbody\n";

        bt_transform->setRotation(rotation);

        motionState = new btDefaultMotionState(*bt_transform);

        if (mass > 0) {
            shape->calculateLocalInertia(mass, inertia);
        }

        btRigidBody::btRigidBodyConstructionInfo rbInfo(mass, motionState,
                                                        shape, inertia);
        rbInfo.m_restitution = restitution;
        rbInfo.m_friction = friction;
        body = new btRigidBody(rbInfo);
        // set user data
        body->setUserPointer(&ID);
        // set trigger
        body->setCollisionFlags(
            body->getCollisionFlags() |
            btCollisionObject::CF_CUSTOM_MATERIAL_CALLBACK);
        if (trigger) {
            body->setCollisionFlags(
                body->getCollisionFlags() |
                btCollisionObject::CO_GHOST_OBJECT);

            body->setCollisionFlags(
                body->getCollisionFlags() |
                btCollisionObject::CF_CUSTOM_MATERIAL_CALLBACK);

            body = (btRigidBody *)(new btGhostObject());
        }

        if (fixedRotation) {
            body->setAngularFactor(btVector3(0, 0, 0));
        }

        Physics3D::instance->dynamicsWorld->addRigidBody(body, 1, 1);
    }

    void Rigidbody3D::AddForce(const glm::vec3 &force) {
        body->applyCentralForce(btVector3(force.x, force.y, force.z));
    }

    void Rigidbody3D::AddTorque(const glm::vec3 &torque) {
        body->applyTorque(btVector3(torque.x, torque.y, torque.z));
    }

    void Rigidbody3D::AddForceAtPosition(const glm::vec3 &force, const glm::vec3 &position) {
        body->applyForce(btVector3(force.x, force.y, force.z),
                         btVector3(position.x, position.y, position.z));
    }

    // velocity functions
    void Rigidbody3D::SetVelocity(const glm::vec3 &velocity) {
        body->setLinearVelocity(
            btVector3(velocity.x, velocity.y, velocity.z));
    }

    void Rigidbody3D::SetAngularVelocity(const glm::vec3 &velocity) {
        body->setAngularVelocity(
            btVector3(velocity.x, velocity.y, velocity.z));
    }

    glm::vec3 Rigidbody3D::GetVelocity() {
        auto velocity = body->getLinearVelocity();
        return glm::vec3(velocity.x(), velocity.y(), velocity.z());
    }

    glm::vec3 Rigidbody3D::GetAngularVelocity() {
        auto velocity = body->getAngularVelocity();
        return glm::vec3(velocity.x(), velocity.y(), velocity.z());
    }

    void Rigidbody3D::Update() {
        if (!body)
            return;

        if (!Runtime::instance->isRunning) return;

        btTransform btTrans = body->getWorldTransform();
        glm::mat4 mat = glm::mat4(1.0f);
        btTrans.getOpenGLMatrix(&mat[0][0]);

        // decompose
        glm::vec3 pos, rot, scal;
        DecomposeTransform(mat, pos, rot, scal);

        transform->position = pos;
        // if (!fixedRotation) {
        transform->rotation = rot;
        // }

        if (fixedRotation) {
            body->setAngularFactor(btVector3(0, 0, 0));
        }
    }

} // namespace Engine::Components
  // namespace Engine::Components